### ISSUE TITLE
Backups over multiple volumes

### ISSUE THREAD

**Issue by hypnotoad** [borgbackup/borg#4896]
I use borg backup for several years now and recommend it to everyone. There is just one thing missing and I am wondering if there is a plan of the borg developer team.

A fundamental limit of borg is that it is located to a single position in your file system and to a single backup medium. So when the backup medium is full, it has to be replaced by a bigger one. Instead, I think it should be possible to just have another volume. E.g., "volume 1" has all the files backuped in 2011-2018 and "volume 2" has 2019-2022 (but none of volume 1).

As far as I see it (reading the documentation several times), it is not possible to achieve this right now. I would assume it should be possible in the following way: When initializing a new volume, the database is copied from the previous volume with additional info that the data is on another volume. The data iteself is not copied.

Are there any plans like that by to borg team?

**Comment by fantasya-pbem** [comment]
Are you aware that even if just have backed up a new file in 2019, the chunks for restoring it may be saved in the volume from 2011, because of deduplication? So restoring just a single file, regardless of it's creating date, could require access to all your volumes, which would be very impractical from my point of view.

**Comment by ThomasWaldmann** [comment]
If you had many backup volumes, you could get the "windows 3.0" install feeling. :-)

(the stuff was on many floppy disks back then and it requested random disks while installing, repeatedly)

**Comment by hypnotoad** [comment]
@fantasya-pbem : Sure, I am aware that all volumes are needed. The volumes are only needed during a restore, so it is not really unpractical.

As main use case, I see "Neverending backups": You simply never delete old backups and start a new volume when a disk is full. That would be practical for almost everyone. If you re-organize your photo files, deduplication kicks in. Of course, you can manually split your backups into something like "all photos" and "all videos", but then you need to do 2 backups with 2 meda every time you do a backup.

Let's not discuss that it would be unpractical because of xyz. I would propose that we keep the thread open in case someone needs the feature or wants to implement it. If there was an FAQ entry "are multi-volume backups possible?", I would not have started this thread.

**Comment by Stunner** [comment]
I agree with the OP (@hypnotoad ). I too have encountered such scenarios and agree it would  be a nice to have feature. That being said, relying on multiple volumes isn't as safe on relying on just 1. Assuming a volume is just a HDD, you increase the chances of your backup being irrecoverable by adding more HDDs as you increase the chance of a failure in the backup array. I use borg within NAS however, so this risk is mitigated by my NAS's RAID array.

**Comment by mo-han** [comment]
right now borg can't do it.
but I do successfully did this once - the repo segments are already incomplete but I created a new archive towards it, the key is that the borg cache helped me with that. but still, this means nothing, it's not useful.

for guys who needs to do incremental backups even the old backups are offline, there is another tool called "DAR backup", though which can only do file-level-dedup rather than block-level-dedup.

**Comment by fuzzdk** [comment]
I think this would be very useful. Especially if you could set up on how many redundant copies in a set of volumes you want to have of your files. This way you could have a set of hdds in a remote location and them sometime take the oldest one home and add new files. It would also be good with a a way to store files where the number of redundant copies is already high enough, just before you are going to move a hdd to the remote location.

**Comment by carlbeech** [comment]
Hi
I wouldn't think that having borg backup over more than one location would be a major issue... however, I was thinking that instead of swapping between backup media, you'd simply just plug in more than one backup USB hard drive...

Linux is perfectly capable of having more than one USB HDD plugged in - all that happens is that you mount them in two mount points - in which case, there should be no reason why borg cannot extend the backup across both drives - so backup chunks that do not fit in disk 1 are simply written to disk 2.

I've got a 2Tb backup disk, and a number of 500Gb disks - I'd really not want to have to go out and buy a 3 or 4 Tb disk when my 2Tb fills up - its costly and a little frustrating if I've got perfectly good disks sitting by the side of me...

Yes, its perfectly possible to have one repository on each disk, and run two backups one to each disk - and splitting the source information across the two backups - however, that means re-working the backups and losing the historical information, as opposed to a graceful expansion...

Anyway - just my 10c worth - please do not think I'm being critical here - I believe borg is one of the best backup mechanisms out there - its just that I think the ability to extend repos across more than one disk would be a very worthwhile addition..

All the best
Carl

**Comment by amerlyq** [comment]
@carlbeech you know, you can create RAID over multiple USB drives. 
Or, more flexibly, add all your USB drives to LVM group and get "one big virtual storage" -- and access it as such after inserting all disks.
Or course, if ANY disk fails -- in most cases all your backups are gone :)
But take into account how borg deduplication will spread your data chunks all over multiple disks anyway -- so if archiving/fragmenting is implemented poorly (without redundancy) you will lose/corrupt most of your backups even using Borg only.

Therefore **consequences of disk failure are almost the same** -- for LVM or borg-custom multidisk solution. 
You may try LVM if increased risk is worth it.

On general topic: of course, if implemented correctly, redundant copies spread over multiple media are great to have.
But then Borg will become not only backup solution, but also **archiving-library-management** solution.
You may look how complex it may become by trying to setup and fully understand git-annex(1).
It's literally configuration complexity hell, which maybe fits into the brain of only the most prudent and disciplined data-hoarders.

**Comment by Massimo-B** [comment]
As of today, spreading a backup over several disks is still not possible?

**Comment by sat-hub** [comment]
Borg does not handle disks. It just writes a "repository", which is a bunch of some directories and files, to a folder on a file system. This file system has to deliver enough free space. This file system has to be either mounted locally on the system where borg is executed (the client) or is accessed via SSH on a storage server that has installed a Borg binary, too.

**Comment by ThomasWaldmann** [comment]
@sat-hub Hello fellow grumpy cat! 🤣

**Comment by IlmariKu** [comment]
I found here via Googling. Here's what I learned. I have my backups as a 2 USB-drive setups in a remote location and I just want to spread the files over two disks easily.

- LVM's are the first thing that comes up, but even with linear-spreading, it seems it's not so easy to recover the files if one disk becomes damaged. I'd just want a solution, where I can plug-and-play the USB-drive and read files from there, if there's a catastrophy or I need backups.
- I thought about writing some script or sharing different directories to either of the volumes. Didn't like this solution, seemed too hacky.
- MergerFS seems to be the solution. Single-volume for borg to point the backups, but spread over two disks.

**Comment by mo-han** [comment]
mergerfs is good, same for rclone union, and sth similar.

**Comment by ThomasWaldmann** [comment]
borg2 (currently in beta) will use `borgstore` and there is already some basic implementation of `MStore`, which can use multiple `Store`s as backends - either for redundancy or to distribute stuff to multiple places.

`MStore` isn't much practically tested yet though and also the implementation needs more work.

**Comment by Massimo-B** [comment]
This could be a game-changer to be able to backup to a series of older harddisks as cold-backup.
Will there be some redundancy for restoring after 1 drive fails like Raid 5 does?
I thought about Raid 5, but that means I need all disks attached. Cold backup disks I would attach via USB. There are 5 disk enclosures, but what if the count of disks increases and I only have 1 disk after the other attached.

It would be great if borg can backup to a series of disks, having 1 disk redundancy like RAID-5. Is that generally possible?

**Comment by ThomasWaldmann** [comment]
I don't think I'll implement rather complicated stuff (like RAID5, ECC codes, ...) in borgstore's `MStore' - if one wants that, one could also add a borgstore backend for something already implementing such features (like e.g. ceph).

But if it is simple enough, like mirroring or distribution, that's better in scope (and partially already there).

**Comment by Massimo-B** [comment]
When adding more devices to the Store, then the risk to damage will increase a lot without redundancy and would straiten the meaning of a backup. Any layer introducing Raid or distributed devices is always based on having all devices available.

Some background why backups over multiple volumes is highly interesting: None of the deduplicating backup strategies like borg or restic is able to backup into slices for tapes or removable mediums. This is often useful when having small old devices consuming too much energy as raid-cluster but are still good enough for cold-backups. One of the only sophisticated backup projects providing backups to slices is dar: http://dar.linux.free.fr/doc/Features.html
But dar cannot cope with all the snapshots that I have on btrfs. AFAIK borg can deduplicate well enough to keep the snapshots deduplicated as it is in btrfs.

**Comment by ciroiriarte** [comment]
That's like going to the old ways of backup, that may not necessarily be a bad thing :)

Usually the hardest part is the media management part.

The software has to manage a pool of media, being that "My offline disk stash". Label each disk and use them appropriately to know which backup jobs go in which disks (even multiple copies), what data needs to be purged because expiration is due or what jobs cannot share media because they are supposed to be redundant copies. Also, single jobs spanning multiple media.

If that's possible, you could have a stash of 500GB/1TB/4TB/16TB SATA disks to protect your super NAS which is plugged 7*24 should ransomware or fat fingering hit.

**Comment by carlbeech** [comment]
If this ever got taken on, what would be a useful 'nice to have' would be
the ability to stipulate that file 'x' would be duplicated across media...
 e.g. if you knew that file 'x' / folder x etc was super-duper important,
and the backup was split across more than one disk, then those files should
be stored on *both* disks as duplicate copies - which would then give you
some guard against media failure...  e.g. you could say 'minimum 2 copies'
- e.g. if you had three disks, then the file would appear on two out of the
three...   - joys of working with hyperconverged storage..

Hope that makes sense.. ;-)

Cheers
Carl.

**Comment by ciroiriarte** [comment]
For sure, nobody wants to replicate distributed filesystems or RAID logic in a backup solution. In the old days, you had a backup job, let's say: "Super important family pictures" and at the job level you stipulate you need two copies.

The backup solution should send two copies in parallel to two different media (2*tapes, or disk+tape). Other solutions like Commvault run an initial backup operation to the media agent (the backup server) and then there is one or several DASH copies to the additional destinations. 1 copy from the client/source, several copies from backup server to media if needed.

In any case, a backup catalog is required, tracking which job definitions have which files set from which client, and also on which media are stored (disk library Y or tape 000289, or AWS bucket Z, etc)

**Comment by Massimo-B** [comment]
Right, a backup catalogue is required just like https://github.com/Edrusb/DAR does. dar was initially historically created for this usecase. But it can't do proper deduplication, where borg is far more powerful.
But when doing a backup of one big expensive 20TiB disk to a list of old 4TiB disks I have in spare, using borg would require to have all these discs online at the same time, creating a RAID or JBOD. It's possible to have a bunch of SATA-to-USB bridges or there are 5bay SATA-USB-docks available. But when it comes to 8 or more discs this is not optimal.
So for now I would rather think borg is not made for this backup approach to a series of disks (or tapes in the old days), if both is needed, deduplication and catalogues for multi-disks. Any other ideas?

**Comment by hypnotoad** [comment]
Realistically, I think there are only 2 ways to implement this.

One is that the overview on the state of a multi-volume repository is on the newest disk and it can refer to other disks. So for deduplication, incremental backups know about chunks on previous disks and skip them. During a recovery, a special procedure is needed to combine data from multiple disks. To me, it sounds like a bigger and riskier change in borg.

The other would be to build software ontop of borg which would manage independent repositories on different volumes. E.g. it would tell  borg to disregard files older than the creation of a repositority. It could maybe pass borg a list of chunks which are available elsewhere. As every repository would be independent, less changes would be needed and the implementation would be safer by design. The borg code would basically be untouched and file recovery could be tested easier.

**Comment by ThomasWaldmann** [comment]
I am closing this as "not planned for the foreseeable future":
- adds significant complexity to the code
- adds complexity to using borg
- as of 2025, there are 36TB HDDs on the market
- one can use existing technologies (like RAID) to combine multiple HDDs
